Sprint 6 solution

aKorishev opened the pull request:
То что можно считать дополнительым заданием (дополнительный кейс) было реализовано ранее, находится в main.java

aKorishev committed:
Переименовал TaskHistoryManager в HistoryManager
Перенес HistoryManager в отдельный пакет

aKorishev committed:
Добавил реализацию двусвязного списка
Покрыл тестами

AntonPozdeev commented on src/taskmanagers/InMemoryTaskManager.java:
```diff
@@ -1,23 +1,23 @@
 
-    InMemoryTaskManager(TaskHistoryManager taskHistoryManager){
-        this.taskHistoryManager = taskHistoryManager;
+    InMemoryTaskManager(HistoryManager historyManager){
```
1) **_public_** InMemoryTaskManager(HistoryManager historyManager) {
2) лучше будет инициализировать так: HistoryManagerFactory.initInMemoryHistoryManager()

aKorishev replied:
1. Ок. Наверное отвалилось, не помню почему сделал закрытым в пакете.
2.  Ок, но это замечание относится к другому классу, К TaskManagerFactory

AntonPozdeev commented on src/historyManagers/HistoryManager.java:
```diff
@@ -0,0 +1,13 @@
+
+public interface HistoryManager {
+    void add(Task element);
+    void remove(Task element);
```
достаточно передать только id таски

aKorishev replied:
Да, так было. Но новое задание подразумевает метод add(Task). Я решил сделать add и remove с одной сигнатурой и хранить ссылки на Task

AntonPozdeev replied:
так как в ТЗ написано void remove(int id);, то будем придерживаться этой концепции

AntonPozdeev commented on src/taskmanagers/InMemoryTaskManager.java:
```diff
@@ -86,16 +86,22 @@ public void clearAllSubTasks(){
     }
```
надо еще удалять из истории в методах clearAllTasks/clearAllEpics/clearAllSubTasks

AntonPozdeev commented on src/taskmanagers/TaskManagerFactory.java:
```diff
@@ -0,0 +1,11 @@
+public class TaskManagerFactory {
+    public static TaskManager initInMemoryTaskManager(){
+
+        return new InMemoryTaskManager(new InMemoryHistoryManager());
```
вместо new InMemoryHistoryManager() стоит писать так HistoryManagerFactory.initInMemoryHistoryManager()

AntonPozdeev commented on src/historyManagers/InMemoryHistoryLinkedList.java:
```diff
@@ -0,0 +1,104 @@
+public class InMemoryHistoryLinkedList  {
+    private InMemoryHistoryNode lastNode = null;
+
+    private final Map<Integer, InMemoryHistoryNode> indexes = new TreeMap<>();
```
TreeMap для чего? HashMap здесь вполне хватит

aKorishev replied:
Здесь все для О(1). У treemap удаление O(ln n), у hashMap O(n).
Выбор treemap для приближения к O(1).

AntonPozdeev replied:
https://hipravin.github.io/algorithm-complexity/ можно почитать, там написаны сложности алгоритмов

aKorishev replied:
Даже в статье, hashmap стремится к O(log n) при определенных случаях, при худшем случае будет O(n).

AntonPozdeev replied:
здесь простой  id, поэтому HashMap будет эффективнее, также в ТЗ написано 
"Чтобы выполнить условие, создайте стандартную HashMap"

AntonPozdeev commented on src/historyManagers/InMemoryHistoryLinkedList.java:
```diff
@@ -0,0 +1,104 @@
+        }
+    }
+
+    public void remove(Task task){
```
так как сюда придет id, то первой строкой стоит сделать так node =  indexes.remove(taskId);, а следующей строкой проверить, что node != null

aKorishev replied:
Здесь у меня логика, которая избегает удаления из индекса, если id уже есть, чтобы не перестраивать treemap при удалении.

AntonPozdeev commented on src/historyManagers/InMemoryHistoryLinkedList.java:
```diff
@@ -0,0 +1,104 @@
+    private final Map<Integer, InMemoryHistoryNode> indexes = new TreeMap<>();
+
+    public void add (Task task){
+        int taskId = task.getTaskId();
```
первым шагом стоит вызвать метод по удалению таски

aKorishev replied:
Здесь у меня логика, которая избегает удаления из индекса, если id уже есть, чтобы не перестраивать treemap при удалении.

AntonPozdeev commented on src/historyManagers/InMemoryHistoryLinkedList.java:
```diff
@@ -0,0 +1,104 @@
+
+    private final Map<Integer, InMemoryHistoryNode> indexes = new TreeMap<>();
+
+    public void add (Task task){
```
в этом методе всё проще
1) надо добавить в класс first ноду
2) конструируем ноду на основе таски и lastNode, для первой ноды ставим null
2) если first == null, то значит у нас вообще ничего нет, мы присваиваем к first ноду
3) иначе присваиваем к lastNode.next =  ноду c 2 этапа
4) и к последней саму ноду

aKorishev replied:
Не так.
Для задачи получения сложности О(1) и получения задач в обратном порядке хватает treemap и односвязного списка. 
Поэтому здесь firstтnode не нужен, т.к. по нему не нужно вызывать перебор элементов, а за поиск отвечает treemap.
nod'ы имеют ссылки на пред и след. элементы для ускорения удаления.

AntonPozdeev replied:
если будет hashMap, то ничего не надо перестраивать будет

aKorishev replied:
TreeMap Больше подходит к задаче. FirtsNode не нужен, т.к. нет задачи выводить fifo

aKorishev committed:
Добавил добавлению в историю при чтении всех задач
Добавил удалиление из истории при удалении всех задач
Покрыл тестами новые методы

aKorishev committed:
Обновил на HistoryManager.remove(int)

aKorishev committed:
Ухудшил производительность выбором РashMap для организации индекса в historyManagers/InMemoryHistoryLinkedList.java
